### PR TITLE
Perform fewer comparisons in balance fn

### PR DESCRIPTION
**Pull request by robinheghan**
Elm currently calls a function every time you perform comparisons (`<`, `<=`...). This is expensive. Performance is improved by reducing comparisons, and by avoiding unnecessary function calls.

### CODE REVIEW

**Comment by BrianHicks** [comment]
Have you run the benchmarks to see how it actually effects performance?

**Comment by BrianHicks** [comment]
Actually that was too brusque. Let me try again:

This looks awesome, and in addition you've memoized a function call. Hooray, and thanks! I'm curious how the performance changed *on your machine*. I would imagine this would speed up insert and remove the most.

**Comment by robinheghan** [comment]
Yes, it is faster on my machine, otherwise i wouldn't have sent the PR. But I definetly should've posted the benchmark results. Note to self for future PRs. :)

**Comment by BrianHicks** [comment]
I'm working on something to automate that… for future PRs. ;)

I'm running the benchmarks now!
